deepsubtree.go

```diff
@@ -37,12 +37,21 @@ func (dsmst *DeepSparseMerkleSubTree) AddBranch(proof SparseMerkleProof, key []b
 		return &BadProofError{}
 	}
 
+	// Update nodes along branch
 	for _, update := range updates {
 		err := dsmst.ms.Set(update[0], update[1])
 		if err != nil {
 			return err
 		}
 	}
 
+	// Update sibling node
+	if proof.SideNodes != nil && len(proof.SideNodes) > 0 {
+		err := dsmst.ms.Set(proof.SideNodes[0], proof.SiblingData)
```

Comment:
proof.SiblingData is supposed to be nullable. Shouldn't this only update the sibling node if it's not null, and return an error perhaps?

Reply:
Even then, not all AddBranches may need the sibling data for their witnesses, so maybe it shouldn't even return an error. We can just allow deleteWithSideNodes to return the error if a branch was added with missing sibling data that is needed.

Reply:
The PR description gives rationale for why it was chosen to make SiblingData optional, namely ergonomics around having two different proofs but no compile-time type safety. That being said, it users really want to shoot themselves in the foot then we can't stop them, so maybe it's not so bad to make it optional.

Reply:
I definitely think that making SibingData compulsory to make Merkle proofs return true is a bad idea. This is imposing potentially large and unnecessary bandwidth burdens on users of the library, that's only needed for a relatively specific use case. This could basically make the library overheads unacceptable for use cases where leaf data is huge.

Reply:
Ah, that might be where your resistance to this stems. You only need to provide the value of the _node_, not the value of the _leaf_. For internal nodes this is 65 bytes (1 byte domain separator, 32 bytes left, 32 bytes right), for leaf nodes this is...also 65 bytes (1 byte domain separator, 32 bytes path, 32 bytes _leaf hash_)! So the overhead, as mentioned in the description is ~60 bytes (exactly 65 in all cases to be exact). Which at the end of the day, if we're considering ~20-40 levels, is about a 5% increase. Might as well be measurement error.

Hence the desire to simply...always include them. It's tiny overhead and completely removes any cognitive load on the developers to manage which proofs are updatable and which are not, and completely prevents developers shooting themselves in the foot.

Reply:
I see. I'm still very against the idea of forcing the sibling data to be included in proofs, for proofs to be valid. It's still an unacceptable overhead for a use case that most projects probably won't use IMO. This also makes the library non-standard, as the nice thing about SMTs is that their Merkle proofs should be standard binary Merkle tree proofs.

I think there are two options:

1. Add another function called ProveUpdatable that developers who use DSMT should use. If they use that function, then it shouldn't be possible to shoot yourself in the foot, unless they edit the proof struct themselves to null SiblingData, in which case it's fine as it's their risk - they can already shoot themselves in the foot by modifying the struct. However, they could shoot themselves in the foot by providing the wrong kind of proofs to AddBranch.

2. Go back to your original idea of having a different type. Create a SparseMerkleProofUpdatable and SparseCompactMerkleProofUpdatable type, that extends SparseMerkleProof and SparseCompactMerkleProof. Since they extend the types, that shouldn't add too much extra complexity. The sanity checks for those types should also be extended. AddBranch should only then accept SparseCompactMerkleProofUpdatable. However, you have to make new functions such as CompactUpdateProof, etc... which adds some complexity.

What do you think?

Reply:
I'd lean towards the former + good docs on exported functions

Reply:
Fixed in 1989b6f.